File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
# N으로 표현

def solution(N, number):
    dp = [[]]
    
    x = 0
    for i in range(1, 9): # N 1 ~ 8개 사용
        dp.append(set())
        x = (10 * x) + N
        dp[i].add(x)  # N, NN, NNN...
        
        for j in range(i):
            for op1 in dp[j]:
                for op2 in dp[i-j]:
                    dp[i].add(op1 + op2)
                    dp[i].add(op1 - op2)
                    dp[i].add(op1 * op2)
                    if op2 != 0:
                        dp[i].add(op1 // op2)
            
            if number in dp[i]:
                return i
        
    return -1

solution(5, 12)

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        List<Set<Integer>> dp = new ArrayList<>();
        
        // Add empty set at position 0
        dp.add(new HashSet<>());
        
        int x = 0;
        for (int i = 1; i <= 8; i++) {
            // N 1 ~ 8개 사용
            Set<Integer> currentSet = new HashSet<>();
            dp.add(currentSet);
            
            x = (10 * x) + N;
            currentSet.add(x);  // N, NN, NNN...
            
            for (int j = 1; j < i; j++) {
                for (int op1 : dp.get(j)) {
                    for (int op2 : dp.get(i - j)) {
                        currentSet.add(op1 + op2);
                        currentSet.add(op1 - op2);
                        currentSet.add(op1 * op2);
                        if (op2 != 0) {
                            currentSet.add(op1 / op2);
                        }
                    }
                }
            }
            
            if (currentSet.contains(number)) {
                return i;
            }
        }
        
        return -1;
    }
}